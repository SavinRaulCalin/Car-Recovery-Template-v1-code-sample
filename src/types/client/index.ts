//  !##################!  //
//  * ----- Base ----- *  //
//  !##################!  //
type Event = React.KeyboardEvent | React.MouseEvent;
type Generic<T> = T;

//  !####################!  //
//  * ----- Navbar ----- *  //
//  !####################!  //
export type TnavEvent = Event;
export type TnavMobile = { onToggle: (e: Event) => void; isOpen: boolean };

//  !###########################!  //
//  * ----- COMMON - CARD ----- *  //
//  !###########################!  //
export type Tparagraph = {
  paragraph: string;
  href?: never;
  value?: never;
  attr?: never;
};
export type Tphone = {
  href: `${"tel:"}${string}`;
  value: string;
  paragraph?: never;
  attr?: never;
};
export type Temail = {
  href: `${"mailto:"}${string}`;
  value: string;
  paragraph?: never;
  attr?: never;
};
export type Tsocial = {
  href: `${"https:"}${string}`;
  value: string;
  paragraph?: never;
  attr: Tattr;
};
export type Tattr = { target: "_blank"; rel: "noreferrer" };
export type Tvariant = Tparagraph | Tphone | Temail | Tsocial;

export type Tbase = {
  svg: JSX.Element;
  title: { text: string; variant: "h2" | "h3" };
  paragraph: string;
  elClass?: string;
};

//  !##################!  //
//  * ----- FORM ----- *  //
//  !##################!  //
export type TonChange<T = string> = { target: { name: string; value: T } };
export type TonFocus<T = string> = { name: T };
export type Tindexable = { [key: string]: string };
export type TonSubmitEvent = React.FormEvent<HTMLFormElement>;
type TbaseFormElement = {
  label: string;
  name: string;
  value: string;
  error: string;
};
export type Tinput = {
  onChange: React.ChangeEventHandler<HTMLInputElement>;
  onFocus: ({ name }: { name: string }) => void;
  type: Generic<"email" | "password" | "text">;
} & TbaseFormElement;

export type Ttextarea = {
  onChange: (e: React.ChangeEvent<HTMLTextAreaElement>) => void;
  onFocus: ({ name }: { name: string }) => void;
  label: string;
  name: string;
  value: string | undefined | null;
  error: string;
};
