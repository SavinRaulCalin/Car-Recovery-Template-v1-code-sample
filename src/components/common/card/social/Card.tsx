import { FC } from "react";
import classNames from "classnames";

import { Tbase, Tvariant } from "@/types/client";
import { Title, Paragraph, Anchor } from "../../typography";
import { Figure } from "../components";

type Tprops = Pick<Tbase, "svg" | "title" | "elClass"> & Tvariant;

const Card: FC<Tprops> = ({ svg, title: { text, variant }, paragraph, href, value, attr, elClass }) => (
  <li className={classNames("mb-[var(--md)] flex justify-start", { [`${elClass}`]: elClass })}>
    <Figure svg={svg} border />
    <header className="flex flex-col justify-center">
      <Title elClass="mb-[2px] font-medium md:text-sm [&::first-letter]:uppercase" text={text} variant={variant} />
      {href && value && <Anchor href={href} value={value} attr={attr} />}
      {paragraph && <Paragraph elClass="!text-sm !tracking-normal md:!text-xs">{paragraph}</Paragraph>}
    </header>
  </li>
);

export default Card;
