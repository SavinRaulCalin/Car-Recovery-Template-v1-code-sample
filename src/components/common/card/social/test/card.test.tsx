import React from "react";
import { render, screen } from "@testing-library/react";

import Card from "../Card";
import { Demo } from "@/svg";
import { Tbase, Tvariant } from "@/types/client";

type Tprops = Pick<Tbase, "svg" | "title" | "elClass"> & Tvariant;

const props: Pick<Tprops, "svg" | "title" | "elClass"> = { svg: <Demo />, title: { text: "title", variant: "h2" }, elClass: "elClass" };
const textProps: Tprops = { ...props, paragraph: "paragraph" };
const telProps: Tprops = { ...props, href: "tel:+44 744544789", value: "(44) 744 544 789" };
const emailProps: Tprops = { ...props, href: "mailto:test@test.com", value: "test@test.com" };
const linkProps: Tprops = { ...props, href: "https://www.link.com", value: "@link", attr: { rel: "noreferrer", target: "_blank" } };

describe("path: components/common/card/social - Card", () => {
  describe("snapshot", () => {
    it("is a match", () => {
      const { asFragment } = render(<Card {...textProps} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("props", () => {
    it("have the correct image", () => {
      render(<Card {...textProps} />);

      expect(screen.getByRole("img")).toBeInTheDocument();
    });

    it("have the correct title", () => {
      render(<Card {...textProps} />);

      expect(screen.getByRole("heading", { level: 2 })).toHaveTextContent("title");
      expect(screen.getByText("title")).toBeInTheDocument();
    });

    it("have the correct style", () => {
      render(<Card {...textProps} />);

      expect(screen.getByRole("listitem")).toHaveClass("elClass");
    });

    describe("Card is of type text", () => {
      it("have the correct text", () => {
        render(<Card {...textProps} />);

        expect(screen.getByTestId("paragraph")).toHaveTextContent("paragraph");
        expect(screen.getByText("paragraph")).toBeInTheDocument();
      });
    });

    describe("Card is of type `tel`", () => {
      it("have the correct text", () => {
        render(<Card {...telProps} />);

        expect(screen.getByTestId("anchor")).toHaveAttribute("href", "tel:+44 744544789");
        expect(screen.getByTestId("anchor")).toHaveTextContent("(44) 744 544 789");
        expect(screen.getByText("(44) 744 544 789")).toBeInTheDocument();
      });
    });

    describe("Card is of type `email`", () => {
      it("have the correct text", () => {
        render(<Card {...emailProps} />);

        expect(screen.getByTestId("anchor")).toHaveAttribute("href", "mailto:test@test.com");
        expect(screen.getByTestId("anchor")).toHaveTextContent("test@test.com");
        expect(screen.getByText("test@test.com")).toBeInTheDocument();
      });
    });

    describe("Card is of type `link`", () => {
      it("have the correct text", () => {
        render(<Card {...linkProps} />);

        expect(screen.getByTestId("anchor")).toHaveAttribute("href", "https://www.link.com");
        expect(screen.getByTestId("anchor")).toHaveTextContent("@link");
        expect(screen.getByText("@link")).toBeInTheDocument();
      });
    });
  });
});
