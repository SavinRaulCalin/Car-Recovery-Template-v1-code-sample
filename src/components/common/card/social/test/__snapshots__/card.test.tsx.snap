// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`path: components/common/card/social - Card snapshot is a match 1`] = `
<DocumentFragment>
  <li
    class="mb-[var(--md)] flex justify-start elClass"
  >
    <figure
      class="mr-4 flex min-h-[55px] min-w-[55px] items-center justify-center border-[1.5px] border-solid border-blue/10 p-2"
    >
      <svg
        fill="none"
        height="50"
        id="demo"
        role="img"
        viewBox="0 0 24 24"
        width="50"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 12C21 16.9706 16.9706 21 12 21C7.02944 21 3 16.9706 3 12C3 7.02944 7.02944 3 12 3C16.9706 3 21 7.02944 21 12Z"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </figure>
    <header
      class="flex flex-col justify-center"
    >
      <h2
        class="font-second leading-none tracking-wide text-blue mb-[2px] font-medium md:text-sm [&::first-letter]:uppercase"
      >
        title
      </h2>
      <p
        class="text-justify text-base tracking-wide text-blue lg:text-sm !text-sm !tracking-normal md:!text-xs"
        data-testid="paragraph"
      >
        paragraph
      </p>
    </header>
  </li>
</DocumentFragment>
`;
